aptos-move/block-executor/src/lib.rs

```diff
@@ -147,3 +149,56 @@ mod txn_last_input_output;
 #[cfg(test)]
 mod unit_tests;
 pub mod view;
+
+/// In standard BlockSTM, the internal states assume contiguous transactions indices.
+/// To also support sharded execution (where each shard gets a subsequence of the transactions
+/// and possibly waits for transaction results from other shards),
+/// an index mapping is need as an additional input to our existing BlockSTM implementation.
+#[derive(Clone)]
+pub struct IndexMapping {
+    pub indices: Vec<TxnIndex>,
+    /// A TxnIndex -> local position mapping.
+    /// Currently implemented as a `Vec` of size equal to the block size, assuming it's not too large.
+    pub inverses: Vec<usize>,
+    /// A fake index representing the end of the transaction list.
+    pub end_index: TxnIndex,
+}
+
+impl IndexMapping {
+    pub fn new(indices: Vec<TxnIndex>, block_size: usize) -> Self {
+        let mut inverses = vec![usize::MAX; block_size];
+        for (pos, &index) in indices.iter().enumerate() {
+            inverses[index as usize] = pos;
+        }
+        Self {
+            indices,
+            inverses,
+            end_index: TxnIndex::MAX,
+        }
+    }
+
+    pub fn new_unsharded(block_size: usize) -> Self {
+        Self {
+            indices: (0..(block_size as TxnIndex)).collect(),
+            inverses: (0..block_size).collect(),
+            end_index: block_size as TxnIndex,
+        }
+    }
+
+    pub fn next_index(&self, index: TxnIndex) -> TxnIndex {
```

Comment:
Maybe return an option?

Reply:
let's discuss [here](https://github.com/aptos-labs/aptos-core/pull/8596#discussion_r1231457484)

Reply:
@sitalkedia was also suggesting to use `Option<TxnIndex>` instead of a special end index (today it's `num_txns`: a txn index of value `num_txns` means it's point to the end of the txn list). But I found some issue while trying it out. some critical states like `Scheduler::execution_idx`, `Scheduler::commit_state` and many other variables may at some points have the "end index" as its value. Making this type change means there will be a lot of if/match/unwrap code for accessing their values (= a lot more diff chunks for you guys to go through too).

I still think that if today we are heavily using the special end index and that has been battle-tested, defining an equivalent for the non-contiguous case is the simplest + safest way to go. `u32::MAX` and `num_block_size` are the best candidates.

@zekun000 @gelash thoughts?